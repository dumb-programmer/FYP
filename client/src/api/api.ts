export function login(data: { email: string; password: string }) {
  return fetch("http://localhost:3000/login", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
    credentials: "include",
  });
}

export function signup(data: {
  firstName: string;
  lastName: string;
  email: string;
  password: string;
}) {
  return fetch("http://localhost:3000/signup", {
    method: "POST",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
    mode: "cors",
  });
}

export function getMessages(chatId: string, page: number) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function searchMessages(chatId: string, query: string, page: number) {
  return fetch(
    `http://localhost:3000/chats/${chatId}/messages?query=${query}&page=${page}`,
    {
      mode: "cors",
      credentials: "include",
    }
  );
}

export function deleteMessage(chatId: string, messageId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}/messages/${messageId}`, {
    method: "DELETE",
    mode: "cors",
    credentials: "include",
  });
}

export function sendPrompt(data: { prompt: string }, chatId?: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "POST",
    mode: "cors",
    credentials: "include",
    body: JSON.stringify(data),
    headers: {
      "Content-Type": "application/json",
    },
  });
}

export function getChats(page: number) {
  return fetch(`http://localhost:3000/chats?page=${page}`, {
    mode: "cors",
    credentials: "include",
  });
}

export function updateChat(chatId: string, name: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "PATCH",
    body: JSON.stringify({ name }),
    headers: {
      "Content-Type": "application/json",
    },
    credentials: "include",
    mode: "cors",
  });
}

export function deleteChat(chatId: string) {
  return fetch(`http://localhost:3000/chats/${chatId}`, {
    method: "DELETE",
    credentials: "include",
    mode: "cors",
  });
}
