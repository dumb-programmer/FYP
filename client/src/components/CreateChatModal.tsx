import { useForm } from "react-hook-form";
import { Action, Content, Header, Modal } from "./modal";
import { createChat } from "@/api/api";

export default function CreateChatModal({ dialogRef }: { dialogRef: React.MutableRefObject<HTMLDialogElement | null> }) {
    const { handleSubmit, register, reset, formState: { isSubmitting } } = useForm({
        defaultValues: {
            name: "",
            document: ""
        }
    });


    const onSubmit = handleSubmit(async (data) => {
        const formData = new FormData();
        formData.set("name", data.name);
        formData.set("document", data.document[0]);
        const response = await createChat(formData);
        if (response.ok) {
            reset();
        }
    });

    return <Modal dialogRef={dialogRef}>
        <Header>
            <h1 className="text-2xl font-bold">Create Chat</h1>
        </Header>
        <Content>
            <form className="flex flex-col gap-4 mt-4">
                <div className="form-control">
                    <label htmlFor="name">Name</label>
                    <input className="input input-bordered" type="text" {...register("name")} required />
                </div>
                <div className="form-control">
                    <label htmlFor="document">Document</label>
                    <input id="document" className="input file-input -ml-4 mt-2" type="file" {...register("document")} required />
                </div>
            </form>
        </Content>
        <Action>
            <button className="btn btn-ghost" type="button" onClick={() => dialogRef.current?.close()}>Cancel</button>
            <button className={`btn btn-primary${isSubmitting ? " loading loading-spinner" : ""}`} type="submit" onClick={onSubmit}>Create</button>
        </Action>
    </Modal>
}