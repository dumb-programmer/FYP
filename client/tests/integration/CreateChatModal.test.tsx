import React, { useRef } from "react";
import { render, screen } from "@testing-library/react";
import { expect, describe, it, vi, beforeEach, afterEach } from "vitest";
import CreateChatModal from "../../src/components/CreateChatModal";
import userEvent from "@testing-library/user-event";
import { createChat } from "../../src/api/api";

vi.mock("../../src/api/api.ts", async () => ({ ...await vi.importActual("../../src/api/api.ts"), createChat: vi.fn().mockReturnValue({ ok: true }) }))

const CreateChatModalWrapper = () => {
    const ref = useRef<HTMLDialogElement | null>(null);

    return <CreateChatModal dialogRef={ref} />;
};

const renderComponent = () => {
    const renderResult = render(<CreateChatModalWrapper />)

    renderResult.container.querySelector("dialog")?.showModal();

    return renderResult;
};

describe("CreateChatModal", () => {
    beforeEach(() => {
        HTMLDialogElement.prototype.showModal = vi.fn(function () {
            this.style.display = 'block';
            this.setAttribute('open', '');
        });

        HTMLDialogElement.prototype.close = vi.fn(function () {
            this.style.display = 'none';
            this.removeAttribute('open');
        });
    });

    afterEach(() => {
        vi.clearAllMocks()
    })

    it("renders correctly", () => {
        renderComponent();

        const modal = screen.getByRole("dialog");

        expect(modal).toMatchSnapshot();
    });

    it("name input works", async () => {
        renderComponent();

        const nameInput = screen.getByRole("textbox");

        await userEvent.type(nameInput, "chat");

        expect(nameInput.value).toEqual("chat");
    });

    it("file input works", async () => {
        renderComponent();

        const fileInput = screen.getByLabelText(/document/i);

        const file = new File(["content"], "chat.pdf", { type: "application/pdf" })

        await userEvent.upload(fileInput, file);

        expect(fileInput.files[0]).toEqual(file);
        expect(fileInput.files).toHaveLength(1);
    });

    it("name input is required", () => {
        renderComponent();

        const nameInput = screen.getByRole("textbox");

        expect(nameInput).toBeRequired();
    });

    it("file input is required", () => {
        renderComponent();

        const fileInput = screen.getByLabelText(/document/i);

        expect(fileInput).toBeRequired();
    });

    it("cancel button closes the modal", async () => {
        renderComponent();

        const cancelBtn = screen.getByRole("button", { name: /cancel/i });

        await userEvent.click(cancelBtn);

        const dialog = screen.queryByRole("dialog");

        expect(dialog).toBeNull();
        expect(HTMLDialogElement.prototype.close).toHaveBeenCalled();
    });

    it("submit button creates a new chat and closes the modal", async () => {
        renderComponent();

        const nameInput = screen.getByRole("textbox");

        await userEvent.type(nameInput, "chat");

        const fileInput = screen.getByLabelText(/document/i);

        const file = new File(["content"], "chat.pdf", { type: "application/pdf" })

        await userEvent.upload(fileInput, file);

        const submitBtn = screen.getByRole("button", { name: /create/i });

        await userEvent.click(submitBtn);

        expect(createChat).toHaveBeenCalled();

        const formData = new FormData();
        formData.set("name", "chat");
        formData.set("document", file);
        expect(createChat).toHaveBeenCalledWith(formData);
    });
})