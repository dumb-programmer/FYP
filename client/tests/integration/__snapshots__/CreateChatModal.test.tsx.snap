// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreateChatModal > renders correctly 1`] = `
<dialog
  class="modal"
  data-testid="modal"
  id=":r0:"
  open=""
  style="display: block;"
>
  <div
    class="modal-box"
  >
    <h1
      class="text-2xl font-bold"
    >
      Create Chat
    </h1>
    <form
      class="flex flex-col gap-4 mt-4"
    >
      <div
        class="form-control"
      >
        <label
          for="name"
        >
          Name
        </label>
        <input
          class="input input-bordered"
          name="name"
          required=""
          type="text"
        />
      </div>
      <div
        class="form-control"
      >
        <label
          for="document"
        >
          Document
        </label>
        <input
          class="input file-input -ml-4 mt-2"
          id="document"
          name="document"
          required=""
          type="file"
        />
      </div>
    </form>
    <div
      class="modal-action"
    >
      <button
        class="btn btn-ghost"
        type="button"
      >
        Cancel
      </button>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Create
      </button>
    </div>
  </div>
</dialog>
`;
